flux-middle/src/rty/projections.rs

```diff
@@ -67,3 +83,220 @@ impl TypeFolder for WithPredicates {
+                    debug_assert_eq!(src_subst.len(), dst_subst.len());
+                    iter::zip(src_subst, &dst_subst)
+                        .for_each(|(src_arg, dst_arg)| self.infer_from_arg(&src_arg, dst_arg));
+                } else {
+                    bug!("unexpected base_ty")
+                }
+            }
+            ty::TyKind::Array(src, _) => {
+                if let BaseTy::Array(dst, _) = Self::base_ty(dst) {
+                    self.infer_from_ty(&src, &dst)
+                } else {
+                    bug!("unexpected base_ty")
+                }
+            }
+
+            ty::TyKind::Slice(src) => {
+                if let BaseTy::Slice(dst) = Self::base_ty(dst) {
+                    self.infer_from_ty(&src, &dst)
+                } else {
+                    bug!("unexpected base_ty")
+                }
+            }
+            ty::TyKind::Tuple(src_tys) => {
+                if let BaseTy::Tuple(dst_tys) = Self::base_ty(dst) {
+                    debug_assert_eq!(src_tys.len(), dst_tys.len());
+                    iter::zip(src_tys.iter(), dst_tys.iter())
+                        .for_each(|(src, dst)| self.infer_from_ty(&src, dst))
+                } else {
+                    bug!("unexpected base_ty")
+                }
+            }
+            _ => {}
+        }
+    }
+}
+
+fn get_impl_source<'tcx>(
+    genv: &GlobalEnv<'_, 'tcx>,
+    elem: DefId,
+    impl_rty: &Ty,
+    callsite_def_id: DefId,
+) -> ImplSourceUserDefinedData<'tcx, Obligation<'tcx, rustc_middle::ty::Predicate<'tcx>>> {
+    // 1a. build up the `Obligation` query
+    let trait_def_id = genv.tcx.parent(elem);
+    let predicate = TraitPredicate {
+        trait_ref: TraitRef::new(genv.tcx, trait_def_id, vec![into_rustc_ty(&genv.tcx, impl_rty)]),
+        constness: rustc_middle::ty::BoundConstness::NotConst,
+        polarity: rustc_middle::ty::ImplPolarity::Positive,
+    };
+    let oblig = Obligation {
+        cause: ObligationCause::dummy(), // TODO(RJ): use with_span instead of `dummy`
+        param_env: genv.tcx.param_env(callsite_def_id),
+        predicate,
+        recursion_depth: 5, // TODO(RJ): made up a random number!
+    };
+
+    // 1b. build up the `SelectionContext`
+    let inf_ctxt = genv.tcx.infer_ctxt().build();
+    let mut sel_ctxt = SelectionContext::new(&inf_ctxt);
+
+    // 1c. issue query to find the `impl` block that implements the `Trait`
+    let impl_source = match sel_ctxt.select(&oblig) {
+        Ok(Some(rustc_middle::traits::ImplSource::UserDefined(impl_source))) => impl_source,
+        Ok(e) => bug!("invalid selection for {oblig:?} = {e:?}"),
+        Err(e) => bug!("error selecting {oblig:?}: {e:?}"),
+    };
+    impl_source
+}
+
+/// Given an an `impl_rty` e.g. `std::vec::IntoIter<Nat>` and an `elem` e.g. `std::iter::Iterator::Item`,
+/// returns the component of the `impl_rty` that corresponds to the `elem`, e.g. `Nat`.
+
+pub fn resolve_impl_projection(
```

Comment:
Instead of having a separate `resolve_impl_projection` this should be integrated into `normalize_projections`. I think like rustc we need to have an equivalent to `normalize_projection_type` (https://doc.rust-lang.org/nightly/nightly-rustc/rustc_trait_selection/traits/project/fn.normalize_projection_type.html) that takes an `AliasTy` (+ all the necessary context) and returns the normalized type. That function first looks for a candidate in `ParamEnv` and if it can't find one then it looks into impl sources (see this function to see how rustc looks for candidates in different places https://doc.rust-lang.org/nightly/nightly-rustc/src/rustc_trait_selection/traits/project.rs.html#1463)